### PR TITLE
snappy,systemd,wrappers: move service units generation to wrappers

### PR DESCRIPTION
**Pull request by pedronis**
This moves service units generation in the wrappers package with some cleanups and porting tests.

In the process found out that systemd Stop was in practice rounding timeouts to 30s, fixed and in the end changed the code to use timers as suggested by Gustavo.

Also some followup to the previous branch to make things more uniform.

### CODE REVIEW

File: snap/info.go [hunk 1]
```diff
@@ -249,10 +249,25 @@ func (app *AppInfo) WrapperPath() string {
 	return filepath.Join(dirs.SnapBinariesDir, binName)
 }
 
+func (app *AppInfo) invocation(command string) string {
+	securityTag := app.SecurityTag()
+	return fmt.Sprintf("/usr/bin/ubuntu-core-launcher %s %s %s", securityTag, securityTag, filepath.Join(app.Snap.MountDir(), command))
```
Comment (zyga) [inline snap/info.go]: Ah, interesting. What is the motivation behind this change? I don't disagree with it in any way, just curious.

File: wrappers/binaries.go [hunk 2]
```diff
@@ -92,7 +93,10 @@ export HOME="$SNAP_USER_DATA"
 	}
 	wrapperData.EnvVars = strings.Join(envVars, "\n")
 
-	t.Execute(&templateOut, wrapperData)
+	if err := t.Execute(&templateOut, wrapperData); err != nil {
+		// this can never happen, except we forget a variable
+		logger.Panicf("Unable to execute template: %v", err)
```
Comment (zyga) [inline wrappers/binaries.go]: +1

File: wrappers/services_test.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	sysdLog = nil
+
+	svcFName := "snap.wat.wat.service"
+
+	err = wrappers.RemoveSnapServices(info, &progress.NullProgress{})
+	c.Assert(err, IsNil)
+
+	c.Check(sysdLog[1], DeepEquals, []string{"stop", svcFName})
+
+	// kill
```
Comment (zyga) [inline wrappers/services_test.go]: Nice! :)

File: snap/info.go (hunk 1 above)
Comment (pedronis) [inline snap/info.go]: the systemd units seems to want the full abs path, it seems sane/safe in general

**Comment by zyga** [comment]
LGTM

File: snap/info.go [hunk 4]
```diff
@@ -249,10 +249,25 @@ func (app *AppInfo) WrapperPath() string {
...
+	return app.invocation(app.Command)
+}
+
+// StopInvocation returns the launcher command line to use when invoking the app stop command binary.
+func (app *AppInfo) StopInvocation() string {
+	return app.invocation(app.Stop)
+}
+
+// PostStopInvocation returns the launcher command line to use when invoking the app post-stop command binary.
+func (app *AppInfo) PostStopInvocation() string {
```
Comment (niemeyer) [inline snap/info.go]: Can we use the following names:
* Command
* StopCommand
* PostStopCommand
* LauncherCommand
* LauncherStopCommand
* LauncherPostStopCommand

These seem to make their meaning and cross-distinction more evident.

Okay to be on a follow up.

Comment (pedronis) [inline snap/info.go]: can do, still in this one I can think

Comment (pedronis) [inline snap/info.go]: @niemeyer done

File: systemd/systemd.go [hunk 5]
```diff
@@ -282,9 +282,21 @@ func (s *systemd) Stop(serviceName string, timeout time.Duration) error {
...
+		steps = toutSteps
+	}
+	if steps < 1 {
+		steps = 1
+	}
+
 	for time.Now().Before(max) {
 		s.reporter.Notify(fmt.Sprintf("Waiting for %s to stop.", serviceName))
-		for i := 0; i < stopSteps; i++ {
+		for i := 0; i < steps; i++ {
```
Comment (niemeyer) [inline systemd/systemd.go]: The logic above feels tricky to read. It took me a while to understand the bug you were fixing, and how it was supposed to work, and then whether it was actually working.

This kind of logic is always more clear if we think about it in terms of time rather than steps. I suggest replacing both of these loops by a single loop based on timers:
```
notify := time.NewTicker(notifyDelay)
check := time.NewTicker(checkDelay)
giveup := time.After(timeout)
for {
    select {
    case <-check:
            ...
    case <-notify:
            ...
    case <-giveup:
            return &Timeout{...}
    }
}
panic("unreachable")

Comment (niemeyer) [inline systemd/systemd.go]: Fixed to defer the timer stops.

File: wrappers/export_test.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+var (
+	// binaries
+	GenerateSnapBinaryWrapper = generateSnapBinaryWrapper
+
+	// services
+	GenerateSnapServiceFile = generateSnapServiceFile
+	GenerateSnapSocketFile  = generateSnapSocketFile
+)
+
+func MockKillWait(wait time.Duration) func() {
```
Comment (niemeyer) [inline wrappers/export_test.go]: Let's please always name these results as `(restore func())`. Makes the idea more self-documenting.

File: wrappers/services.go [hunk 7]
```diff
@@ -199,18 +197,18 @@ Requires=snapd.frameworks.target{{ if .App.Socket }} {{.SocketFileName}}{{end}}
 X-Snappy=yes
 
 [Service]
-ExecStart=/usr/bin/ubuntu-core-launcher {{.App.SecurityTag}} {{.App.SecurityTag}} {{.FullPathStart}}
+ExecStart={{.App.LauncherCommand}}
```
Comment (niemeyer) [inline wrappers/services.go]: Thanks! The template is much more clear now.

**Comment by niemeyer** [comment]
One more note, and LGTM!

File: systemd/systemd.go [hunk 8]
```diff
@@ -280,24 +282,34 @@ func (s *systemd) Stop(serviceName string, timeout time.Duration) error {
...
-		s.reporter.Notify(fmt.Sprintf("Waiting for %s to stop.", serviceName))
-		for i := 0; i < stopSteps; i++ {
+	giveup := time.NewTimer(timeout)
+	notify := time.NewTicker(stopNotifyDelay)
+	defer notify.Stop()
+	check := time.NewTicker(stopCheckDelay)
+	defer check.Stop()
+
+	firstCheck := true
+loop:
```
Comment (mvo5) [inline systemd/systemd.go]: I like this new way of doing this loop

**Comment by mvo5** [comment]
This looks fine, thanks! One thing I noticed is that the new `wrappers/services.go` has a `interacter` that has `Agreed` and `Notify`. I think we just need a `notifier` with `Notify` here. But that is not a blocker.
